### PR TITLE
[4] Move check for offline mode from render to dispatch so that modules are not rendered on 403

### LINKED ISSUE #29455: Frontend redirect fails after the site is set as "offline
### Steps to reproduce the issue
After setup the frontend as "offline", the following url is not redirected to frontend offline page:
http://sitename.com/index.php/component/config/

### Expected result

Site Offline page only.

### Actual result
The website is showing the frontend template along with: the 403 error in content position, displaying modules like: Older Posts, Most read posts, and the main menu module. 

### System information (as much as possible)

Default Joomla 4.0.0 beta1 package with sample data installed.

### PR DESCRIPTION
**Pull request by PhilETaylor**
Closes https://github.com/joomla/joomla-cms/issues/29455

### Summary of Changes

Move check for offline mode from `render` to `dispatch` method so that modules are not rendered on 403

### Testing Instructions

Set your Joomla 4 site to offline mode in Joomla global config

Visit the frontend. 

Attempt to login as admin - you can - great - you can see the site - great. 

Attempt to go to https://example.com/index.php/component/config = 403 Error (Some module positions will render like headers and footers and maybe menus depending on your template 

Apply PR, repeat test, this time at https://example.com/index.php/component/config you will get a login box. 

### Actual result BEFORE applying this Pull Request

Attempt to go to https://example.com/index.php/component/config = 403 Error (Some module positions will render like headers and footers and maybe menus depending on your template 

### Expected result AFTER applying this Pull Request

this time at https://example.com/index.php/component/config you will get a login box. You can login to see your site

### Also

Cleans up two unused language vars.

### Documentation Changes Required

None

### CODE REVIEW

File: libraries/src/Application/SiteApplication.php [hunk 1]
```diff
@@ -154,9 +154,6 @@ public function dispatch($component = null)
 		switch ($document->getType())
 		{
 			case 'html':
-				// Get language
```
Comment (PhilETaylor) [inline libraries/src/Application/SiteApplication.php]: These three lines are unused in this class and are safe to delete (I think!)

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 3964cc975d5c5dc5244f99c9b133aff5aa9e45cd<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33879">issues.joomla.org/tracker/joomla-cms/33879</a>.</sub>

**Comment by ceford** [comment]
With or without the patch I don't get 403 - just the Home page (multilingual site).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33879">issues.joomla.org/tracker/joomla-cms/33879</a>.</sub>

**Comment by PhilETaylor** [comment]
@ceford what url did you test to get the 403?

**Comment by ceford** [comment]
localhost/j4test/index.php/component/config/

but reloads as http://localhost/j4test/index.php/en/

also used https - Macbook Pro with Cataline / PHP8 / ... merged latest commits early this morning.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33879">issues.joomla.org/tracker/joomla-cms/33879</a>.</sub>

**Comment by PhilETaylor** [comment]
this just means your language redirects are kicking in before this PRs code, which is not a bad thing

What if you start at http://localhost/j4test/index.php/en/component/config/ - please run your testing starting at that url.

**Comment by ceford** [comment]
http://localhost/j4test/index.php/en/component/config/ gives me a 403 page with Banner, Top menu and Footer with Login. I also get an error call stack:

     403 You don't have permission to access this. Please contact a website administrator if this is incorrect. 
 JROOT/components/com_config/src/Dispatcher/Dispatcher.php:50 

with and without the patch applied<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33879">issues.joomla.org/tracker/joomla-cms/33879</a>.</sub>

**Comment by ceford** [comment]
That last comment was wrong - I forgot to go off-line again. Actual result is as before.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33879">issues.joomla.org/tracker/joomla-cms/33879</a>.</sub>

**Comment by PhilETaylor** [comment]
So what you are saying is that yours was a successful test, when a site is multi-lingual, that you cannot replicate the issue this PR resolves, and your testing showed that this PR did not make your experience, with a multi-lingual site any worse.

**Comment by ceford** [comment]
Yes - I cannot reproduce the problem described on my multi-lingual test site. I do not get a 403 page.

**Comment by ghost** [comment]
The [successfully test](https://github.com/joomla/joomla-cms/pull/33879#issuecomment-841610960) above was on a multilanguage site. 

I've been test now using index.php/de/component/config (+ fr, fa), always successfully.

**Comment by PhilETaylor** [comment]
DO NOT TEST on a multi-lingual site. That never had the issue to start with, so this PR doesn't solve anything on a multi-lingual site. 

All your tests prove on a multi-lingual site is that I did not break anything further. 

Please try to replicate the reported issue. Replicate the issue, Apply the PR, repeat, and see the issue is resolved.

File: libraries/src/Application/SiteApplication.php [hunk 2]
```diff
@@ -181,8 +192,6 @@ public function dispatch($component = null)
 					$wr->addTemplateRegistryFile($template->parent, $this->getClientId());
 				}
 
-				$wr->addTemplateRegistryFile($template->template, $this->getClientId());
```
Comment (Fedik) [inline libraries/src/Application/SiteApplication.php:182]: And here you just broke an assets override feature :)

Now the template asset cannot override assets from a components.

Comment (PhilETaylor) [reply libraries/src/Application/SiteApplication.php:184]: Its offline mode, there is no component rendering... right after this we return, and we dont render any component.

**Comment by ghost** [comment]
> Please try to replicate the reported issue. Replicate the issue, Apply the PR, repeat, and see the issue is resolved.

That's what i have done. 

I was not aware to not test on a multilang site, sorry.

File: libraries/src/Application/SiteApplication.php (hunk 2 above)
Comment (PhilETaylor) [reply libraries/src/Application/SiteApplication.php:184]: oh I see, you want this line back... got it...

Comment (PhilETaylor) [reply libraries/src/Application/SiteApplication.php:182]: Actually I think you missed the fact that, although this diff shows this line deleted, what you missed is that its not deleted. Its moved from line 184 to line 171 (before rendering the offline.php) and therefore it has not broken anything :)

Comment (Fedik) [reply libraries/src/Application/SiteApplication.php:182]: yes, I seen it is moved, and this is  broke the override

Comment (Fedik) [reply libraries/src/Application/SiteApplication.php:182]: order is important here

Comment (PhilETaylor) [reply libraries/src/Application/SiteApplication.php:182]: I have reordered the flow now - please see if that fixes is.

Comment (Fedik) [reply libraries/src/Application/SiteApplication.php:182]: that is better, thanks

Comment (PhilETaylor) [reply libraries/src/Application/SiteApplication.php:182]: Thank you for testing and the feedback. Good catch.

**Comment by PhilETaylor** [comment]
> I was not aware to not test on a multilang site, sorry.

You can test on a multi-lang site, but you will need to replicate the issue on that site. To do that you will need a URL like this:

`http://example.com/index.php/en/component/config`

with the /en/ part (where en is one of your languages)

**Comment by ghost** [comment]
I had the issue on the multilanguage site. That's the first thing i do: do i understand the issue, can i replicate it. Then i test the pull request, the test was successfully.

**Comment by PhilETaylor** [comment]
ah sorry - this is why I need coffee before reading emails :)

**Comment by chmst** [comment]
@PhilETaylor could you please resolve conflicts?

**Comment by PhilETaylor** [comment]
> @PhilETaylor could you please resolve conflicts?

There are no conflicts that I can see to resolve. I have merged the upstream branch for the second time.

**Pull request closed without merging**